### PR TITLE
add API paddle.linalg.lu/lu_unpack

### PR DESCRIPTION
**Pull request by zhiboniu**
### PR types

New features
### PR changes

APIs
### Describe

add API paddle.linalg.lu

拆成三个小的子pr合入：
lu_op正向：
https://github.com/PaddlePaddle/Paddle/pull/38559

lu_op反向：
https://github.com/PaddlePaddle/Paddle/pull/38616

lu API/lu_unpack op/API:
https://github.com/PaddlePaddle/Paddle/pull/38617

            import paddle 

            x = paddle.to_tensor([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]]).astype('float64')
            lu,p,info = paddle.linalg.lu(x, get_infos=True)

            '''
            >>> lu:
            Tensor(shape=[3, 2], dtype=float64, place=CUDAPlace(0), stop_gradient=True,
                [[5.        , 6.        ],
                    [0.20000000, 0.80000000],
                    [0.60000000, 0.50000000]])
            >>> p
            Tensor(shape=[2], dtype=int32, place=CUDAPlace(0), stop_gradient=True,
                [3, 3])
            >>> info
            Tensor(shape=[], dtype=int32, place=CUDAPlace(0), stop_gradient=True,
                0)
            '''

            P,L,U = paddle.linalg.lu_unpack(lu,p)

            '''
            >>> P
            (Tensor(shape=[3, 3], dtype=float64, place=CUDAPlace(0), stop_gradient=True,
            [[0., 1., 0.],
            [0., 0., 1.],
            [1., 0., 0.]]), 
            >>> L
            Tensor(shape=[3, 2], dtype=float64, place=CUDAPlace(0), stop_gradient=True,
            [[1.        , 0.        ],
            [0.20000000, 1.        ],
            [0.60000000, 0.50000000]]), 
            >>> U
            Tensor(shape=[2, 2], dtype=float64, place=CUDAPlace(0), stop_gradient=True,
            [[5.        , 6.        ],
            [0.        , 0.80000000]]))
            '''

image
image

image

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 1]
```diff
@@ -1570,6 +1571,8 @@ def check_grad_with_place(self,
             fp32_numeric_grads.append(grad)
         numeric_grads = fp32_numeric_grads
 
+        print(numeric_grads)
+        print(analytic_grads)
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/op_test.py]: delete print?

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 2]
```diff
@@ -1585,6 +1588,8 @@ def check_grad_with_place(self,
                     max_relative_error = 0.03 if max_relative_error < 0.03 else max_relative_error
                 fp32_grads.append(grad)
             dygraph_grad = fp32_grads
+            print(numeric_grads)
+            print(dygraph_grad)
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/op_test.py]: delete print?

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 1 above)
Comment (zhiboniu) [reply python/paddle/fluid/tests/unittests/op_test.py]: done

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 2 above)
Comment (zhiboniu) [reply python/paddle/fluid/tests/unittests/op_test.py]: done

File: paddle/fluid/platform/dynload/lapack.h [hunk 3]
```diff
@@ -23,6 +23,9 @@ limitations under the License. */
 // Note(zhouwei): because lapack doesn't provide appropriate header file.
 // should expose API statement yourself.
 
+#define LAPACK_ROW_MAJOR 101
+#define LAPACK_COL_MAJOR 102
```
Comment (jeff41404) [inline paddle/fluid/platform/dynload/lapack.h]: these 2 macros are not used in any other files

File: python/paddle/fluid/tests/unittests/test_cholesky_op.py [hunk 4]
```diff
@@ -80,74 +80,70 @@ def init_config(self):
...
+#                                       [16, 17, 18]]]).astype("float64")
+#                 input = fluid.dygraph.to_variable(input_np)
+#                 try:
+#                     result = paddle.cholesky(input)
+#                 except RuntimeError as ex:
+#                     print("The mat is singular")
+#                     pass
+#                 except ValueError as ex:
+#                     print("The mat is singular")
+#                     pass
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/test_cholesky_op.py]: these annotation are expected?

File: python/paddle/tensor/linalg.py [hunk 5]
```diff
@@ -1711,6 +1711,213 @@ def qr(x, mode="reduced", name=None):
...
+            [[1.        , 0.        ],
+            [0.20000000, 1.        ],
+            [0.60000000, 0.50000000]]), 
+            >>> U
+            Tensor(shape=[2, 2], dtype=float64, place=CUDAPlace(0), stop_gradient=True,
+            [[5.        , 6.        ],
+            [0.        , 0.80000000]]))
+            '''
+
+            # one can verify : X = P * L * U ;
```
Comment (jeff41404) [inline python/paddle/tensor/linalg.py]: use matmul instead of \* may be more clear

Comment (zhiboniu) [reply python/paddle/tensor/linalg.py]: done

File: python/paddle/fluid/tests/unittests/test_cholesky_op.py (hunk 4 above)
Comment (zhiboniu) [reply python/paddle/fluid/tests/unittests/test_cholesky_op.py]: done

File: paddle/fluid/platform/dynload/lapack.h (hunk 3 above)
Comment (zhiboniu) [reply paddle/fluid/platform/dynload/lapack.h]: done

**Review by TCChenlong — APPROVED** [review]
LGTM

**Pull request closed without merging**